// @toolkip/binding/dist/binder.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const object_helpers_1 = require("@toolkip/object-helpers");
const async_1 = require("@toolkip/async");
class _Binder {
    constructor() {
        this._id = 0;
        this._started = false;
        this._boundDetails = {};
        this._startAnimationLoop();
    }
    _getNextId() {
        this._id += 1;
        return this._id.toString();
    }
    bind(evalFunc, updateFunc, skipFunc, equalsFunc) {
        if (!evalFunc || !updateFunc) {
            return "";
        }
        let lastValue = evalFunc();
        if (lastValue) {
            async_1.nextRender().then(() => updateFunc(lastValue));
        }
        let details = {
            id: this._getNextId(),
            eval: evalFunc,
            update: (val) => {
                updateFunc(val);
            },
            skip: skipFunc || (() => false),
            lastValue: lastValue,
            equals: equalsFunc || ((a, b) => { return a === b; })
        };
        this._boundDetails[details.id] = details;
        return details.id;
    }
    unbind(key) {
        if (!this._boundDetails[key]) {
            return false;
        }
        delete this._boundDetails[key];
        return true;
    }
    _startAnimationLoop() {
        if (this._started) {
            return;
        }
        this._onFrame();
    }
    _onFrame() {
        return __awaiter(this, void 0, void 0, function* () {
            object_helpers_1.map(this._boundDetails, (details) => {
                this._handlingBinding(details);
            });
            return async_1.nextRender().then(() => this._onFrame());
        });
    }
    _handlingBinding(details) {
        return __awaiter(this, void 0, void 0, function* () {
            if (details.skip()) {
                return;
            }
            let newVal = details.eval();
            if (details.equals(newVal, details.lastValue)) {
                return;
            }
            details.lastValue = newVal;
            details.update(newVal);
        });
    }
}
exports.Binder = new _Binder();
