// @toolkip/primitive-helpers/dist/dates.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const numbers_1 = require("./numbers");
const strings_1 = require("./strings");
function dateDiff(a, b, signed, includeTime, returnMilli) {
    let ms;
    let diff;
    let dir;
    ms = (1000 * 60 * 60 * 24);
    if (!includeTime) {
        a = clearTimeInfo(a, true);
        b = clearTimeInfo(b, true);
    }
    if ((a > b) || signed) {
        diff = (a - b);
    }
    else {
        diff = (b - a);
    }
    if (!returnMilli) {
        diff = diff / ms;
    }
    return diff;
}
exports.dateDiff = dateDiff;
;
var InclusivityEnum;
(function (InclusivityEnum) {
    InclusivityEnum[InclusivityEnum["EXCLUSIVE"] = -1] = "EXCLUSIVE";
    InclusivityEnum[InclusivityEnum["DEFAULT"] = 0] = "DEFAULT";
    InclusivityEnum[InclusivityEnum["INCLUSIVE"] = 1] = "INCLUSIVE";
})(InclusivityEnum = exports.InclusivityEnum || (exports.InclusivityEnum = {}));
function monthDiff(a, b, signed, inclusivity) {
    let monthDiff;
    let yearDiff;
    if ((a > b) || signed) {
        monthDiff = (a.getMonth()) - (b.getMonth());
        yearDiff = (a.getFullYear()) - (b.getFullYear());
    }
    else {
        monthDiff = (b.getMonth()) - (a.getMonth());
        yearDiff = (b.getFullYear()) - (a.getFullYear());
    }
    let diff = yearDiff * 12 + monthDiff;
    diff += +inclusivity;
    return diff;
}
exports.monthDiff = monthDiff;
function getToday(include_time) {
    ;
    let ret;
    ret = new Date();
    if (include_time)
        return ret;
    ret = clearTimeInfo(ret);
    return ret;
}
exports.getToday = getToday;
;
function clearTimeInfo(date, clearTZ) {
    let dateStr = shortDate(date);
    let outDate;
    if (clearTZ) {
        outDate = new Date(dateStr + " 00:00Z");
    }
    else {
        outDate = new Date(dateStr);
    }
    return outDate;
}
exports.clearTimeInfo = clearTimeInfo;
function businessDateDiff(a, b, signed, includeTime, returnMilli) {
    ;
    let diff;
    let dayOfWeek;
    let dir;
    let idx;
    diff = dateDiff(a, b, signed, includeTime, returnMilli);
    dayOfWeek = (b > a ? a.getDay() : b.getDay()) + 1;
    dayOfWeek %= 7;
    if (dayOfWeek < 0) {
        dayOfWeek = 6;
    }
    let weekendDays = 0;
    for (idx = 0; idx < Math.abs(diff); idx += 1) {
        if (dayOfWeek === 0 || dayOfWeek === 6) {
            weekendDays += 1;
        }
        dayOfWeek += 1;
        dayOfWeek %= 7;
        if (dayOfWeek < 0) {
            dayOfWeek = 6;
        }
    }
    if (diff < 0) {
        dir = -1;
    }
    else {
        dir = 1;
    }
    return diff - (weekendDays * dir);
}
exports.businessDateDiff = businessDateDiff;
;
function shortDate(dt) {
    ;
    if (!dt) {
        return "";
    }
    let yr;
    yr = getShortYear(dt);
    return (dt.getMonth() + 1) + "/" + dt.getDate() + "/" + yr;
}
exports.shortDate = shortDate;
;
function inputDateFmt(dt) {
    ;
    let m;
    let d;
    let y;
    y = dt.getFullYear();
    m = (dt.getMonth() + 1);
    if (m < 10)
        m = "0" + m;
    d = +dt.getDate();
    if (d < 10)
        d = "0" + d;
    return (dt.getFullYear() + "-" + m + "-" + d);
}
exports.inputDateFmt = inputDateFmt;
;
function inputToDate(iDt, iTime) {
    let outDate;
    if (iDt) {
        let dtArr = iDt.split("-");
        outDate = new Date(+dtArr[0], +dtArr[1] - 1, +dtArr[2]);
    }
    else if (iTime) {
        outDate = getToday();
    }
    else {
        outDate = null;
        return outDate;
    }
    if (iTime) {
        let timeArr = iTime.split(":");
        outDate.setHours(+timeArr[0]);
        outDate.setMinutes(+timeArr[1]);
    }
    return outDate;
}
exports.inputToDate = inputToDate;
;
function shortTime(dt, withExtra) {
    ;
    let min;
    let min_str;
    let hours;
    let half;
    min = +dt.getMinutes();
    hours = +dt.getHours();
    half = "";
    if (min < 10) {
        min_str = "0" + min;
    }
    else {
        min_str = min.toString();
    }
    if (withExtra) {
        half = " AM";
        if (hours >= 12)
            half = " PM";
        if (hours > 12)
            hours -= 12;
        if (hours === 0) {
            hours = 12;
        }
    }
    return hours + ":" + min_str + half;
}
exports.shortTime = shortTime;
;
function inputTimeFmt(time, includeSeconds) {
    let out = [];
    let hours = time.getHours();
    out.push(numbers_1.padToDigits(hours, 2));
    let minutes = time.getMinutes();
    out.push(numbers_1.padToDigits(minutes, 2));
    if (includeSeconds) {
        let seconds = time.getSeconds();
        out.push(numbers_1.padToDigits(seconds, 2));
    }
    return out.join(":");
}
exports.inputTimeFmt = inputTimeFmt;
function shortDateTime(dt, with_extra) {
    return shortDate(dt) + " " + shortTime(dt, with_extra);
}
exports.shortDateTime = shortDateTime;
;
function stopwatchDisplay(milli, noLeadingZeros, noBlanks) {
    let seconds;
    let minutes;
    let hours;
    let days;
    let arr;
    let sec_str;
    let min_str;
    let hr_str;
    if (!noLeadingZeros) {
        sec_str = strings_1.addLeadingZeroes(2, seconds);
        min_str = strings_1.addLeadingZeroes(2, minutes);
        hr_str = strings_1.addLeadingZeroes(2, hours);
    }
    else {
        sec_str = seconds.toString();
        min_str = minutes.toString();
        hr_str = hours.toString();
    }
    return days + "D  " + hr_str + ":" + min_str + ":" + sec_str + " '" + milli;
}
exports.stopwatchDisplay = stopwatchDisplay;
;
function _retrieveCountsFromMilli(milli) {
    let out = {};
    let remaining = milli;
    out.days = Math.floor(remaining / (24 * 60 * 60 * 1000));
    remaining -= (out.days * 24 * 60 * 60 * 1000);
    out.hours = Math.floor(remaining / (60 * 60 * 1000));
    remaining -= (out.hours * 60 * 60 * 1000);
    out.minutes = Math.floor(remaining / (60 * 1000));
    remaining -= (out.minutes * 60 * 1000);
    out.seconds = Math.floor(remaining / 1000);
    remaining -= (out.seconds * 1000);
    out.milliseconds = remaining;
    return out;
}
function updatedStopwatchDisplay(milli, options) {
    let diffs = _retrieveCountsFromMilli(milli);
    let out = [];
    if (diffs.days) {
        out.push(diffs.days + " days");
    }
    if (diffs.hours) {
        out.push(diffs.hours + " hours");
    }
    if (diffs.minutes) {
        out.push(diffs.minutes + " minutes");
    }
    if (diffs.seconds) {
        out.push(diffs.seconds + " seconds");
    }
    if (diffs.milliseconds && options.showMilli) {
        out.push(diffs.milliseconds + " ms");
    }
    return out.join(" ");
}
exports.updatedStopwatchDisplay = updatedStopwatchDisplay;
function addToDate(date, counts) {
    if (counts.milliseconds) {
        date.setMilliseconds(date.getMilliseconds() + counts.milliseconds);
    }
    if (counts.seconds) {
        date.setSeconds(date.getSeconds() + counts.seconds);
    }
    if (counts.minutes) {
        date.setMinutes(date.getMinutes() + counts.minutes);
    }
    if (counts.hours) {
        date.setHours(date.getHours() + counts.hours);
    }
    if (counts.days) {
        date.setDate(date.getDate() + counts.days);
    }
    if (counts.months) {
        date.setMonth(date.getMonth() + counts.months);
    }
    if (counts.years) {
        date.setFullYear(date.getFullYear() + counts.years);
    }
    return date;
}
exports.addToDate = addToDate;
;
function getMonthName(date, short) {
    switch (date.getMonth()) {
        case 0:
            if (short)
                return "Jan";
            return "January";
        case 1:
            if (short)
                return "Feb";
            return "February";
        case 2:
            if (short)
                return "Mar";
            return "March";
        case 3:
            if (short)
                return "Apr";
            return "April";
        case 4:
            return "May";
        case 5:
            if (short)
                return "Jun";
            return "June";
        case 6:
            if (short)
                return "Jul";
            return "July";
        case 7:
            if (short)
                return "Aug";
            return "August";
        case 8:
            if (short)
                return "Sept";
            return "September";
        case 9:
            if (short)
                return "Oct";
            return "October";
        case 10:
            if (short)
                return "Nov";
            return "November";
        case 11:
            if (short)
                return "Dec";
            return "December";
    }
    return "";
}
exports.getMonthName = getMonthName;
;
function getDayOfWeek(date, short) {
    ;
    switch (date.getDay()) {
        case 0:
            if (short)
                return "Sun";
            return "Sunday";
        case 1:
            if (short)
                return "Mon";
            return "Monday";
        case 2:
            if (short)
                return "Tues";
            return "Tuesday";
        case 3:
            if (short)
                return "Wed";
            return "Wednesday";
        case 4:
            if (short)
                return "Thurs";
            return "Thursday";
        case 5:
            if (short)
                return "Fri";
            return "Friday";
        case 6:
            if (short)
                return "Sat";
            return "Saturday";
    }
    return "";
}
exports.getDayOfWeek = getDayOfWeek;
;
function getLengthOfMonthInDays(date) {
    if (!date) {
        return -1;
    }
    let month = date.getMonth();
    switch (month) {
        case 0:
        case 2:
        case 4:
        case 6:
        case 7:
        case 9:
        case 11:
            return 31;
        case 1:
            if (isLeapYear(date)) {
                return 29;
            }
            else {
                return 28;
            }
        default:
            return 30;
    }
}
exports.getLengthOfMonthInDays = getLengthOfMonthInDays;
function isLeapYear(date) {
    if (!date) {
        return false;
    }
    let year = date.getFullYear();
    if (year % 4 !== 0) {
        return false;
    }
    if ((year % 100 === 0) && (year % 400 !== 0)) {
        return false;
    }
    return true;
}
exports.isLeapYear = isLeapYear;
function getShortYear(date) {
    return (+date.getFullYear() % 100);
}
exports.getShortYear = getShortYear;
function isWeekend(date) {
    let dayOfWeek = date.getDay();
    if (dayOfWeek === 0) {
        return true;
    }
    if (dayOfWeek === 6) {
        return true;
    }
    return false;
}
exports.isWeekend = isWeekend;
function isToday(date) {
    let today = getToday();
    let cloneDate = clearTimeInfo(date);
    return isSameDate(today, cloneDate);
}
exports.isToday = isToday;
function isSameDate(dateA, dateB) {
    if (shortDate(dateA) === shortDate(dateB)) {
        return true;
    }
    return false;
}
exports.isSameDate = isSameDate;
function getDisplayDuration(counts) {
    _updateDateDifferences(1000, counts, "milliseconds", "seconds");
    _updateDateDifferences(60, counts, "seconds", "minutes");
    _updateDateDifferences(60, counts, "minutes", "hours");
    _updateDateDifferences(24, counts, "hours", "days");
    _updateDateDifferences(30, counts, "days", "months");
    _updateDateDifferences(12, counts, "months", "years");
    let out = [];
    if (counts.years) {
        out.push(_createPluralString(counts.years, "year"));
    }
    if (counts.months) {
        out.push(_createPluralString(counts.months, "month"));
    }
    if (counts.days) {
        out.push(_createPluralString(counts.days, "day"));
    }
    if (counts.hours) {
        out.push(_createPluralString(counts.hours, "hour"));
    }
    if (counts.minutes) {
        out.push(_createPluralString(counts.minutes, "minute"));
    }
    if (counts.seconds) {
        out.push(_createPluralString(counts.seconds, "second"));
    }
    if (counts.milliseconds) {
        out.push(_createPluralString(counts.milliseconds, "millisecond"));
    }
    return out.join(" ");
}
exports.getDisplayDuration = getDisplayDuration;
function _updateDateDifferences(divisor, out, startKey, endKey) {
    if (!out[startKey]) {
        out[startKey] = 0;
    }
    if (!out[endKey]) {
        out[endKey] = 0;
    }
    let dividend = out[startKey];
    let remainder = dividend % divisor;
    let quotient = Math.floor(dividend / divisor);
    out[startKey] = remainder;
    out[endKey] += quotient;
    return out;
}
function _createPluralString(amount, singular, plural) {
    if (amount === 1) {
        return amount + " " + singular;
    }
    else {
        if (!plural) {
            plural = singular + "s";
        }
        return amount + " " + plural;
    }
}
