// @toolkip/maths/dist/rectConversion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _1 = require(".");
function clientRectToShape(rect) {
    let out;
    out = new Array();
    out[0] = {
        x: rect.left,
        y: rect.top
    };
    out[1] = {
        x: rect.left + rect.width,
        y: rect.top
    };
    out[2] = {
        x: rect.left + rect.width,
        y: rect.top + rect.height
    };
    out[3] = {
        x: rect.left,
        y: rect.top + rect.height
    };
    return out;
}
exports.clientRectToShape = clientRectToShape;
function svgRectToShape(rect) {
    let out;
    out = new Array();
    out[0] = {
        x: rect.x,
        y: rect.y
    };
    out[1] = {
        x: rect.x + rect.width,
        y: rect.y
    };
    out[2] = {
        x: rect.x + rect.width,
        y: rect.y + rect.height
    };
    out[3] = {
        x: rect.x,
        y: rect.y + rect.height
    };
    return out;
}
exports.svgRectToShape = svgRectToShape;
function svgRectToBasicRect(rect) {
    let out;
    out = {
        x: rect.x,
        y: rect.y,
        w: rect.width,
        h: rect.height
    };
    return out;
}
exports.svgRectToBasicRect = svgRectToBasicRect;
;
function clientRectToBasicRect(rect) {
    let out;
    out = {
        x: rect.left + window.scrollX,
        y: rect.top + window.scrollY,
        w: rect.width,
        h: rect.height
    };
    return out;
}
exports.clientRectToBasicRect = clientRectToBasicRect;
function toBasicRect(rect) {
    let r;
    if (_1.isIBasicRect(rect)) {
        r = rect;
    }
    else if (_1.isClientRect(rect)) {
        r = clientRectToBasicRect(rect);
    }
    else if (_1.isSVGRect(rect)) {
        r = svgRectToBasicRect(rect);
    }
    return r;
}
exports.toBasicRect = toBasicRect;
;
