// @toolkip/maths/dist/trig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const maths_1 = require("./maths");
function degreesToRadians(deg) {
    let result = ((Math.PI * deg) / 180);
    return result;
}
exports.degreesToRadians = degreesToRadians;
function radiansToDegrees(rad) {
    let result = (180 * rad) / Math.PI;
    return result;
}
exports.radiansToDegrees = radiansToDegrees;
function getEndPoint(startPoint, deg, distance) {
    let rad = degreesToRadians(deg);
    let result = {
        x: (Math.cos(rad) * distance) + startPoint.x,
        y: (Math.sin(rad) * distance) + startPoint.y
    };
    return result;
}
exports.getEndPoint = getEndPoint;
function getCentralPoint(elem) {
    return {
        x: elem.offsetLeft + (elem.offsetWidth / 2),
        y: elem.offsetTop + (elem.offsetWidth / 2)
    };
}
exports.getCentralPoint = getCentralPoint;
function getAngle(start, end) {
    ;
    let dx;
    let dy;
    let q_sign;
    let q_ang;
    let angle;
    dx = (end.x - start.x);
    dy = (end.y - start.y);
    if (dx === 0)
        return (dy < 0) ? 270 : 90;
    if (dy === 0)
        return (dx < 0) ? 180 : 0;
    q_sign = ((dx * dy) > 0) ? 1 : -1;
    q_ang = (dx < 0) ? Math.PI : 0;
    angle = Math.atan(Math.abs(dy) / Math.abs(dx));
    angle = ((angle * q_sign) + q_ang);
    return (angle * (180 / Math.PI));
}
exports.getAngle = getAngle;
;
function getDistance(start, end) {
    ;
    let distance;
    let dx;
    let dy;
    dx = (start.x - end.x);
    dy = (start.y - end.y);
    distance = Math.sqrt((dx * dx) + (dy * dy));
    return distance;
}
exports.getDistance = getDistance;
;
function calculatePolygonInternalAngle(numberOfSides) {
    return maths_1.roundToPlace(degreesToRadians(360 / numberOfSides), 1000);
}
exports.calculatePolygonInternalAngle = calculatePolygonInternalAngle;
