// @toolkip/style-libraries/dist/__tests__/placeholderLibrary.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = require("..");
const style_helpers_1 = require("@toolkip/style-helpers");
const stylesA = {
    ".one": {
        backgroundColor: "<bg>",
        fontSize: "<size>"
    },
    ".two": {
        color: "<bg>",
        width: "<width>"
    }
};
const stylesB = {
    ".any": {
        fontSize: "<size>"
    },
    ".two": {
        fontSize: "<size>"
    }
};
describe('placeholderLibrary', () => {
    beforeAll(() => {
        __1.PlaceholderLibrary.add("A", stylesA);
        __1.PlaceholderLibrary.add("B", stylesB);
    });
    beforeEach(() => {
        __1.PlaceholderLibrary.replacePlaceholder({
            newValue: "<size>",
            placeholder: "size",
        });
    });
    it("indexes styles by placeholder", () => {
        __1.PlaceholderLibrary.add("test", stylesB);
        expect(__1.PlaceholderLibrary["_indexedPlaceholders"]).toMatchObject({
            size: {
                test: {
                    ".any": {
                        fontSize: true
                    },
                    ".two": {
                        fontSize: true
                    }
                }
            }
        });
    });
    it("merges styles with the same placeholder", () => {
        expect(__1.PlaceholderLibrary["_indexedPlaceholders"].size).toMatchObject({
            A: {
                ".one": {
                    fontSize: true
                }
            },
            B: {
                ".any": {
                    fontSize: true
                },
                ".two": {
                    fontSize: true
                }
            }
        });
    });
    it("creates elements per placeholder + unique key", () => {
        expect(__1.PlaceholderLibrary["_elems"]).toHaveProperty("A-bg");
        expect(__1.PlaceholderLibrary["_elems"]).toHaveProperty("A-size");
        expect(__1.PlaceholderLibrary["_elems"]).toHaveProperty("A-width");
    });
    it("replaces placeholders globally", () => {
        __1.PlaceholderLibrary.replacePlaceholder({
            newValue: "10rem",
            placeholder: "size"
        });
        expect(__1.PlaceholderLibrary["_elems"]["A-size"].innerHTML)
            .toEqual(style_helpers_1.stringifyStyles({ ".one": { fontSize: "10rem" } })[0]);
        expect(__1.PlaceholderLibrary["_elems"]["B-size"].innerHTML)
            .toEqual(style_helpers_1.stringifyStyles({
            ".any": { fontSize: "10rem" },
            ".two": { fontSize: "10rem" }
        })[0]);
    });
    it("replaces placeholders within a unique key", () => {
        __1.PlaceholderLibrary.replacePlaceholder({
            newValue: "5rem",
            placeholder: "size",
            uniqueKey: "A"
        });
        expect(__1.PlaceholderLibrary["_elems"]["A-size"].innerHTML)
            .toEqual(style_helpers_1.stringifyStyles({ ".one": { fontSize: "5rem" } })[0]);
        expect(__1.PlaceholderLibrary["_elems"]["B-size"].innerHTML)
            .toEqual(style_helpers_1.stringifyStyles({
            ".any": { fontSize: "<size>" },
            ".two": { fontSize: "<size>" }
        })[0]);
    });
});
