// @toolkip/model/dist/_typeguards/specific.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const identifiableModel_1 = require("../objectModels/identifiableModel");
const objectModels_1 = require("../objectModels");
const primitiveModels_1 = require("../primitiveModels");
const arrayModels_1 = require("../arrayModels");
const selectors_1 = require("../helpers/selectors");
exports.isPrimitiveModel = (test) => {
    if (test instanceof primitiveModels_1.MPrimitive) {
        return true;
    }
    return false;
};
exports.isDateModel = (test) => {
    if (test instanceof primitiveModels_1.MDate) {
        return true;
    }
    return false;
};
exports.isObjectModel = (test) => {
    if (test instanceof objectModels_1.MObject) {
        return true;
    }
    return false;
};
exports.isArrayModel = (test) => {
    if (test instanceof arrayModels_1.MArray) {
        return true;
    }
    return false;
};
exports.isManagerModel = (test) => {
    if (test instanceof arrayModels_1.MManager) {
        return true;
    }
    return false;
};
exports.isIdentifiableModel = (model) => {
    if (model instanceof identifiableModel_1.MIdentifiable) {
        return true;
    }
    return false;
};
exports.isSelector = (test) => {
    if (test instanceof selectors_1.Selector) {
        return true;
    }
    return false;
};
